src/format/opamFile.ml

```diff
@@ -1913,9 +1913,11 @@ module URLSyntax = struct
            (Pp.V.string -| Pp.of_module "checksum" (module OpamHash)));
       "mirrors", Pp.ppacc with_mirrors mirrors
         (Pp.V.map_list ~depth:1 Pp.V.url);
+      (* Disable for 2.1.0
```

Comment:
> I believe this should still be present

If ti is present, we'll have parse error if it is present, or we'll need to update opam file version.
Here, what is disabled is only the parsing printing of the supbath as an url field, not the internal representation of the field itself, which is populated by the `x-subpath` field. Linting is already checked (5bbf2fb64ae) for `x-subpath`. The idea is to keep it with only `x-subpath` field for next releases, and when a new version of the opam file can be released, at that time update opam file & linting enabling the `subpath` field itself ans its checks

Reply:
Ah, I see. My idea was to already have the code for handling files with `opam-version: 2.1`, just not use that by default. I believe it can make the transition smoother.
- if `opam-version < 2.1`
  * on parsing, `x-subpath` is handled, and `subpath` is rejected
  * on printing, always use `x-subpath` (anyway we need to discriminate on opam-version in this case, since the field is stored in the same way in the record)
- if `opam-version >= 2.1`
  * on parsing, both are handled, but `x-subpath` is a lint warning
  * on printing, always use `subpath`

Reply:
One thought - if opam-repository is remaining at `2.0` and therefore `opam` files in repos probably want to remain compatible with `2.0`, should we be considering the `subpath` bit itself as part of `opam-version: "3.0"` (i.e. bumping the major version of the file format at some point in the future and dropping all collected `x-` fields in the process?). i.e. **opam 2.1** supports this, but **file format** 3.0 is what will support it?

Reply:
yeah, we were actually discussing it with @rjbou: there is no need for the file format versions to be synchronised with the opam version (and in fact, there are only loosely, e.g. a new opam version doesn't need to upgrade the file format version). Jumping to 3.0 indeed would avoid the ambiguity so I like the idea. The name of the field becomes slightly misleading, and one or two error messages (advising an upgrade of opam to version xx) should be fixed, but that's the only downside.

As for my position vs. @rjbou's: I now understand more clearly what differs. We have two ways to update:
- hard format change, as for 1.2→2.0 : an opam 1.2 file is no longer a valid opam 2.0 file, and the repository needs to be converted. Opam can still read files in the older format, but by going through a rewrite step first
- soft format change: the new format is an extension, files in the older format are still valid files in the new format.

We just added a field here, so that can fit in the second case: we don't need a rewrite to upgrade opam-repository. We _could_ hint that with a minor format version bump;

The addition of the compatibility layer through `x-foo` doesn't actually need to change that:
- when writing, encode using the new field or the `x-foo` field depending on the file format version (and add _or remove_ the availability constraint on the opam version)
- keep the old version when creating and writing files by default, for the moment.

I don't think it can hurt to have full support for files with the new `subpath` field in the code already, as long as it's not the format used by default.

Reply:
Yes, we can have it from now integrated in the opam file syntax. What was disturbing me from having it handled from now, is the ambiguity that we could have with an file-format at `2.1` and the corresponding opam version `2.2` or more. Increasing directly the file-format version to 3.0 is a good compromise (we will have at some point file format at 4.0 and opam at 3.0).
The _problem_ here is that we don't have the information of the `opam-version` field when we define parsing/printing url field.
For clarity: note that for the moment (in this pr), the `subpath` and `x-subpath` fields are only generated and used by/in opam internals, it is not intended to be used or manipulated by user (the only user interaction is at an `opam pin edit`)

Reply:
I have to begin every GitHub comment at the moment with "sorry for the lag" 🙂

I agree that's unfortunate that we end up `opam-version` the field meaning the format of the file and `opam-version` the variable meaning the version of opam, but I guess the sooner that gets properly uncoupled in people's minds, the better! You already can't use the opam-version *field* to be say, `2.0.2` in order to force the package to be only readable by a maintenance release which has a bug fix you know matters. That already belongs in the available field.

We can of course have the information for `opam-version` available in the parser and lexer in future - I already in https://github.com/ocaml/opam/pull/3350/commits/78d06a8e4c0ff8fd02f4f4d6e4ef9de0a2aef55a assumed (incorrectly) that `opam-version` should be the first field. It might be sensible either to enforce that from now, or at least have `opam lint` warn if the `opam-version` field appears too late in a file.

I think my last concern with this - is still an error, therefore, for a user to provide a file which has `opam-version: "2.1"` in it? i.e. can we formally have the notion that the `2.1` is strictly internal to opam, because there are no fields requiring it for user-provided files?